net/server.go

```diff
@@ -447,3 +461,136 @@ func (s *server) SendPubSubMessage(
+		return true
+	}
+
+	rawblock, err := s.peer.blockstore.Get(s.peer.ctx, c)
+	if err != nil {
+		log.ErrorE("Failed to get block", err)
+		return false
+	}
+	block, err := coreblock.GetFromBytes(rawblock.RawData())
+	if err != nil {
+		log.ErrorE("Failed to get doc from block", err)
+		return false
+	}
+
+	cols, err := s.peer.db.GetCollections(
+		s.peer.ctx,
+		client.CollectionFetchOptions{
+			SchemaVersionID: immutable.Some(block.Delta.GetSchemaVersionID()),
+		},
+	)
+	if err != nil {
+		log.ErrorE("Failed to get collections", err)
+		return false
+	}
+	if len(cols) == 0 {
+		log.Info("No collections found", corelog.Any("Schema Version ID", block.Delta.GetSchemaVersionID()))
+		return false
+	}
+
+	// If the requesting peer is in the replicators list for that collection, then they have access.
+	if peerList, ok := s.replicators[cols[0].SchemaRoot()]; ok {
+		_, exists := peerList[p]
+		if exists {
+			return true
+		}
+	}
+
+	identFunc := func() immutable.Option[identity.Identity] {
+		s.piMux.RLock()
+		ident, ok := s.peerIdentities[p]
+		s.piMux.RUnlock()
+		if !ok {
+			resp, err := s.getIdentity(s.peer.ctx, p)
+			if err != nil {
+				log.ErrorE("Failed to get identity", err)
+				return immutable.None[identity.Identity]()
+			}
+			ident, err = identity.FromToken(resp.IdentityToken)
+			if err != nil {
+				log.ErrorE("Failed to parse identity token", err)
+				return immutable.None[identity.Identity]()
+			}
+			err = identity.VerifyAuthToken(ident, s.peer.PeerID().String())
+			if err != nil {
+				log.ErrorE("Failed to verify auth token", err)
+				return immutable.None[identity.Identity]()
+			}
+			s.piMux.Lock()
+			s.peerIdentities[p] = ident
+			s.piMux.Unlock()
+		}
+		return immutable.Some(ident)
+	}
+
+	peerHasAccess, err := permission.CheckDocAccessWithIdentityFunc(
+		s.peer.ctx,
+		identFunc,
+		s.peer.acp.Value(),
+		cols[0], // For now we assume there is only one collection.
+		acp.ReadPermission,
+		string(block.Delta.GetDocID()),
+	)
+	if err != nil {
+		log.ErrorE("Failed to check access", err)
+		return false
+	}
+
+	return peerHasAccess
+}
+
+// trySelfHasAccess checks if the local node has access to the given block.
+//
+// This is a best-effort check and returns true unless we explicitly find that we don't have access.
+func (s *server) trySelfHasAccess(block *coreblock.Block) bool {
+	if !s.peer.acp.HasValue() {
+		return true
+	}
+
+	cols, err := s.peer.db.GetCollections(
+		s.peer.ctx,
+		client.CollectionFetchOptions{
+			SchemaVersionID: immutable.Some(block.Delta.GetSchemaVersionID()),
+		},
+	)
+	if err != nil {
+		log.ErrorE("Failed to get collections", err)
+		return true
+	}
+	if len(cols) == 0 {
+		log.Info("No collections found", corelog.Any("Schema Version ID", block.Delta.GetSchemaVersionID()))
+		return true
+	}
+	ident, err := s.peer.db.GetNodeIdentity(s.peer.ctx)
+	if err != nil {
+		log.ErrorE("Failed to get node identity", err)
+		return true
+	}
+	if !ident.HasValue() {
```

Comment:
question: Is `ident` the identity of the node that may or may not be set? i.e. not in any way the client's requesting identity correct?

Reply:
Five lines above :)
```
ident, err := s.peer.db.GetNodeIdentity(s.peer.ctx)
```